core/dbt/node_runners.py

```diff
@@ -492,3 +493,66 @@ def print_result_line(self, result):
                                               schema_name,
                                               self.node_index,
                                               self.num_nodes)
+
+
+class RPCCompileRunner(CompileRunner):
+    def __init__(self, config, adapter, node, node_index, num_nodes):
+        super(RPCCompileRunner, self).__init__(config, adapter, node,
+                                               node_index, num_nodes)
+
+    def before_execute(self):
+        pass
+
+    def after_execute(self, result):
+        pass
+
+    def compile(self, manifest):
+        return compile_node(self.adapter, self.config, self.node, manifest, {},
+                            write=False)
+
+    def execute(self, compiled_node, manifest):
+        return RemoteCompileResult(
+            raw_sql=compiled_node.raw_sql,
+            compiled_sql=compiled_node.injected_sql
+        )
+
+    def error_result(self, node, error, start_time, timing_info):
+        raise dbt.exceptions.RPCException(error)
+
+    def ephemeral_result(self, node, start_time, timing_info):
+        raise dbt.exceptions.NotImplementedException(
+            'cannot execute ephemeral nodes remotely!'
+        )
+
+    def from_run_result(self, result, start_time, timing_info):
+        timing = [t.serialize() for t in timing_info]
+        return RemoteCompileResult(
+            raw_sql=result.raw_sql,
+            compiled_sql=result.compiled_sql,
+            timing=timing
+        )
+
+
+class RPCExecuteRunner(RPCCompileRunner):
+    def from_run_result(self, result, start_time, timing_info):
+        timing = [t.serialize() for t in timing_info]
+        return RemoteRunResult(
+            raw_sql=result.raw_sql,
+            compiled_sql=result.compiled_sql,
+            table=result.table,
+            timing=timing
+        )
+
+    def execute(self, compiled_node, manifest):
+        status, table = self.adapter.execute(compiled_node.injected_sql,
+                                             fetch=True)
+        table = [
+            {k: v for k, v in zip(table.column_names, row)}
+            for row in table
+        ]
+
+        return RemoteRunResult(
+            raw_sql=compiled_node.raw_sql,
+            compiled_sql=compiled_node.injected_sql,
+            table=table
```

Comment:
yeah, let's add `table.column_names` in here as a list, separate from `table`

Reply:
@beckjake ping on this one. Sorry for fragmenting the discussion across comments. The use case is that a query might return zero rows, but will contain defined columns. Adding a separate `columns` list will help us render a table of results with zero rows in it